File: samples/msgext-northwind-inventory-action-csharp/NorthwindDB/DbService.cs
using Azure;
using Azure.Data.Tables;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using NorthwindInventory.Models;

namespace NorthwindInventory.NorthwindDB
{
    public interface IDbEntity : ITableEntity
    {
        string ETag { get; set; }
        string? PartitionKey { get; set; }
        string? RowKey { get; set; }
        DateTimeOffset? Timestamp { get; set; }
    }

    public interface IDbProject : IDbEntity, IProduct { }
    public interface IDbSupplier : IDbEntity, ISupplier { }

    public class DbService<TEntity> where TEntity : class, IDbEntity, new()
    {
        private readonly bool _okToCacheLocally;
        private List<TEntity> _entityCache = new List<TEntity>();
        private IConfiguration _configuration;

        public DbService(bool okToCacheLocally, IConfiguration configuration)
        {
            _configuration = configuration;
            if (string.IsNullOrEmpty(_configuration["StorageConnectionString"]))
            {
                throw new InvalidOperationException("STORAGE_ACCOUNT_CONNECTION_STRING is not set");
            }
            _okToCacheLocally = okToCacheLocally;
        }

        public async Task<TEntity> GetEntityByRowKeyAsync(string tableName, string rowKey)
        {
            if (!_okToCacheLocally)
            {
                var tableClient = new TableClient(_configuration["StorageConnectionString"], tableName);

                // Ensure you provide the partitionKey
                var partitionKey = "PartitionKey"; // Replace with actual partitionKey or logic to determine it.

                var result = await tableClient.GetEntityAsync<TEntity>(partitionKey, rowKey);
                return ExpandPropertyValues(result);
            }
            else
            {
                var result = await GetEntitiesAsync(tableName);
                var entity = result.FirstOrDefault(e => e.RowKey == rowKey);
                if (entity == null)
                {
                    Console.WriteLine("Not found");
                    return null;
                }
                return entity;
            }
        }

        public async Task<List<TEntity>> GetEntitiesAsync(string tableName)
        {
            if (_okToCacheLocally && _entityCache.Any())
            {
                return _entityCache;
            }
            else
            {
                var tableClient = new TableClient(_configuration["StorageConnectionString"], tableName);
                var entities = new List<TEntity>();

                // Use QueryAsync to get entities
                await foreach (var entity in tableClient.QueryAsync<TEntity>())
                {
                    entities.Add(ExpandPropertyValues(entity));
                }

                _entityCache = entities;
                return _entityCache;
            }
        }

        public async Task CreateEntityAsync(string tableName, string rowKey, TEntity newEntity)
        {
            var tableClient = new TableClient(_configuration["StorageConnectionString"], tableName);
            await tableClient.UpsertEntityAsync(newEntity);            
        }

        public async Task UpdateEntityAsync(TEntity updatedEntity)
        {
            var tableClient = new TableClient(_configuration["StorageConnectionString"], updatedEntity.PartitionKey);
            var etag = new ETag(updatedEntity.ETag);
            await tableClient.UpdateEntityAsync(updatedEntity, etag, TableUpdateMode.Replace);
        }

        public async Task<int> GetNextIdAsync(string tableName)
        {
            try
            {
                var entities = await GetEntitiesAsync(tableName);
                return entities.Count + 1;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error fetching entities: {ex.Message}");
                throw;
            }
        }

        private TEntity ExpandPropertyValues(TEntity entity)
        {
            var result = new TEntity();
            var properties = typeof(TEntity).GetProperties();
            foreach (var prop in properties)
            {
                var value = prop.GetValue(entity);
                prop.SetValue(result, ExpandPropertyValue(value));
            }
            return result;
        }

        private object ExpandPropertyValue(object value)
        {
            if (value is string str && (str.StartsWith('{') || str.StartsWith('[')))
            {
                try
                {
                    return JsonSerializer.Deserialize<object>(str);
                }
                catch
                {
                    return value;
                }
            }
            return value;
        }

        private TEntity CompressPropertyValues(TEntity entity)
        {
            var result = new TEntity();
            var properties = typeof(TEntity).GetProperties();
            foreach (var prop in properties)
            {
                var value = prop.GetValue(entity);
                prop.SetValue(result, CompressPropertyValue(value));
            }
            return result;
        }

        private object CompressPropertyValue(object value)
        {
            if (value is object obj)
            {
                return JsonSerializer.Serialize(obj);
            }
            return value;
        }
    }
}

File: samples/cext-trey-research-auth/src/services/DbService.ts
import { TableClient, TableEntity } from "@azure/data-tables";
import { HttpError } from './Utilities';
import { DbEntity } from '../model/dbModel';

export default class DbService<DbEntityType> {

    private storageAccountConnectionString = process.env.STORAGE_ACCOUNT_CONNECTION_STRING;
    private okToCacheLocally = false;
    private entityCache: DbEntity[] = [];

    constructor(okToCacheLocally: boolean) {
        if (!this.storageAccountConnectionString) {
            throw new Error("STORAGE_ACCOUNT_CONNECTION_STRING is not set");
        }
        this.okToCacheLocally = okToCacheLocally;
    }

    async getEntityByRowKey(tableName: string, rowKey: string): Promise<DbEntity> {
        if (!this.okToCacheLocally) {
            const tableClient = TableClient.fromConnectionString(this.storageAccountConnectionString, tableName);
            const result = this.expandPropertyValues(await tableClient.getEntity(tableName, rowKey) as DbEntityType);
            return result as DbEntity;
        } else {
            let result = await this.getEntities(tableName);
            result = result.filter((e) => {
                return e.rowKey === rowKey
            });
            if (result.length === 0) {
                throw new HttpError(404, `Entity ${rowKey} not found`);
            } else {
                return result[0];
            }
        }
    }

    async getEntities(tableName: string): Promise<DbEntity[]> {

        if (!this.okToCacheLocally || this.entityCache.length === 0) {
            // Rebuild cache for this entity
            const tableClient = TableClient.fromConnectionString(this.storageAccountConnectionString, tableName);
            const entities = tableClient.listEntities();
            this.entityCache = [];
            for await (const entity of entities) {
                // Remove any duplicates which sometimes occur after a watch restart
                if (this.entityCache.find((e) => e.rowKey === entity.rowKey) === undefined) {
                    const e = this.expandPropertyValues(entity as DbEntityType);
                    this.entityCache.push(e as DbEntity);
                }
            }
        }
        return this.entityCache;
    }

    async createEntity(tableName: string, rowKey: string, newEntity: DbEntityType): Promise<void> {

        this.entityCache = [];
        const entity = this.compressPropertyValues(newEntity) as DbEntityType;
        const tableClient = TableClient.fromConnectionString(this.storageAccountConnectionString, tableName);
        try {
             await tableClient.createEntity({
                partitionKey: tableName,
                rowKey,
                ...entity
            });
        } catch (ex) {
            if (ex.response?.status !== 409) {
                throw new HttpError(500, ex.message);
            }
        }
}

    async updateEntity(tableName: string, updatedEntity: DbEntityType): Promise<void> {

        this.entityCache = [];
        const e = this.compressPropertyValues(updatedEntity) as DbEntityType;
        const tableClient = TableClient.fromConnectionString(this.storageAccountConnectionString, tableName);
        await tableClient.updateEntity(e as TableEntity, "Replace");

    }

    private expandPropertyValues(entity: DbEntityType): DbEntityType {
        const result = {} as DbEntityType;
        for (const key in entity) {
            result[key] = this.expandPropertyValue(entity[key]);
        }
        return result;
    }

    private expandPropertyValue(v: any): any {
        if (typeof v === "string" && (v.charAt(0) === '{' || v.charAt(0) === '[')) {
            try {
                return JSON.parse(v);
            }
            catch (e) {
                return v;
            }
        } else {
            return v;
        }
    };

    private compressPropertyValues(entity: DbEntityType): DbEntityType {
        const result = {} as DbEntityType;
        for (const key in entity) {
            result[key] = this.compressPropertyValue(entity[key]);
        }
        return result;
    }

    private compressPropertyValue(v: any): any {
        if (typeof v === "object") {
            return JSON.stringify(v);
        } else {
            return v;
        }
    };

}
